### DISCUSSION TITLE
Centering a widget with a fixed width/height

### DISCUSSION THREAD

**Discussion by jwodder** [ratatui/ratatui#631] [Discussion]
If I have a widget with a display width of `width` that I want to center horizontally and/or vertically within a `Rect`, is there a simple way to do the necessary calculations with just `Layout`?  Currently, I'm doing this:

```rust
fn center_horizontally(width: u16, area: Rect) -> Rect {
    let left_margin = (area.width.saturating_sub(width)) / 2;
    let chunks = Layout::default()
        .direction(Direction::Horizontal)
        .constraints([
            Constraint::Length(left_margin),
            Constraint::Length(width.min(area.width)),
            Constraint::Min(0),
        ])
        .split(area);
    chunks[1]
}
```

but I feel like ratatui should have a way to take care of the `left_margin` calculation for me if I use the proper constraints.  Using `Constraint::Ratio` for the "margins" doesn't do what I want, and none of the other constraints look appropriate.  Is this worth a feature request?

**Comment by joshka** [comment]
Does `[Min(0), Length(width), Min(0)]` do what you need?

**Reply by jwodder** [reply]
No, that results in the first chunk having a width of 0.

**Reply by joshka** [reply]
Oops - I forgot that behavior is not yet available due to https://github.com/ratatui-org/ratatui/issues/536
Come help us decide a good name for this.

**Reply by jwodder** [reply]
To be clear: You're saying that one day I could get what I want with:

```rust
let chunks = Layout::default()
    .direction(Direction::Horizontal)
    .constraints([
        Constraint::Min(0),
        Constraint::Length(width),
        Constraint::Min(0),
    ])
    .segment_size(SegmentSize::EvenDistribution)
    .split(area);
```

I'm assuming for this that `SegmentSize::EvenDistribution` would only affect the variable-size constraints (i.e., not `Length`), though that behavior is not what I'd expect at first.

Personally, the API I'd like to see would be `Constraint::Fill(u16)`, where the `u16` is a "pressure" argument used when multiple `Fill`s are present; e.g., `Fill(2)` is assigned twice as much size as `Fill(1)`.

**Reply by joshka** [reply]
>To be clear: You're saying that one day I could get what I want with:

Yes. I don't think the `segment_size` name really makes it clear what this does though.

**Reply by joshka** [reply]
I made a PR to make segment_size an unstable feature. https://github.com/ratatui-org/ratatui/pull/633

You can test this in your app with an override:
https://doc.rust-lang.org/cargo/reference/overriding-dependencies.html#the-patch-section

> Personally, the API I'd like to see would be `Constraint::Fill(u16)`

If we are going to add features like this, the right move is to move to using Taffy for layout, which supports flex and grid layouts. Otherwise we end up trying to build everything Taffy would bring on top of Cassowary.

See https://github.com/search?q=repo%3Aratatui-org%2Fratatui+taffy&type=issues for some more in depth reading on this.

**Reply by BppleMan** [reply]
the https://github.com/search?q=repo%3Aratatui-org%2Fratatui+taffy&type=issues has nothing

**Reply by joshka** [reply]
The method you want is called flex.

**Reply by joshka** [reply]
Ah, I misunderstood. We changed the org name of ratatui-org to ratatui. So the correct search is https://github.com/search?q=repo%3Aratatui%2Fratatui+taffy&type=issues
